### PR TITLE
core: libmbedtls: CMAC: remove code duplication

### PR DESCRIPTION
**Pull request by uglym8**
@jenswi-linaro , 
this is follow-up PR to fix the issue raised in
https://github.com/OP-TEE/optee_os/pull/4381#discussion_r574309070
@jforissier you might be interested as well as we still have to handle **panic() vs. assert()** stuff I guess.

### CODE REVIEW

File: lib/libmbedtls/core/cmac.c [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+struct mbed_cmac_ctx {
+	struct crypto_mac_ctx mac_ctx;
+	mbedtls_cipher_context_t cipher_ctx;
+};
+
+static const struct crypto_mac_ops mbed_des3_cmac_ops;
+static const struct crypto_mac_ops mbed_aes_cmac_ops;
+
+static struct mbed_cmac_ctx *to_cmac_ctx(struct crypto_mac_ctx *ctx, const mbedtls_cipher_id_t cipher_id)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:28]: Coding style: always try to limit line length to 80 characters, write as many parameters as possible on each line, and prefer alignment on the opening parenthesis (although it is also acceptable to indent less if line length would be exceeded).
```C
static struct mbed_cmac_ctx *to_cmac_ctx(struct crypto_mac_ctx *ctx,
					 const mbedtls_cipher_id_t cipher_id)
```

File: lib/libmbedtls/core/cmac.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		assert(ctx->ops == &mbed_aes_cmac_ops);
+
+	if (cipher_id == MBEDTLS_CIPHER_ID_3DES)
+		assert(ctx->ops == &mbed_des3_cmac_ops);
+
+	return container_of(ctx, struct mbed_cmac_ctx, mac_ctx);
+}
+
+static TEE_Result mbed_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len, const mbedtls_cipher_id_t cipher_id)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:42]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+static TEE_Result mbed_aes_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len)
+{
+	return mbed_cmac_init(ctx, key, len, MBEDTLS_CIPHER_ID_AES);
+}
+
+static TEE_Result mbed_des3_cmac_update(struct crypto_mac_ctx *ctx,
+		const uint8_t *data, size_t len)
+{
+	if (mbedtls_cipher_cmac_update(&to_cmac_ctx(ctx, MBEDTLS_CIPHER_ID_3DES)->cipher_ctx,
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:84]: Use a helper variable

File: lib/libmbedtls/core/cmac.c [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+				data, len))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_aes_cmac_update(struct crypto_mac_ctx *ctx,
+		const uint8_t *data, size_t len)
+{
+	if (mbedtls_cipher_cmac_update(&to_cmac_ctx(ctx, MBEDTLS_CIPHER_ID_AES)->cipher_ctx,
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:94]: Here also, helper variable

File: lib/libmbedtls/core/cmac.c [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+static TEE_Result mbed_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len, const mbedtls_cipher_id_t cipher_id)
+{
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
+	uint8_t block_size = cipher_id == MBEDTLS_CIPHER_ID_AES ? TEE_AES_BLOCK_SIZE : TEE_DES_BLOCK_SIZE;
+	uint8_t block_digest[TEE_AES_BLOCK_SIZE] = { };
+
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:108]: Remove blank line

File: lib/libmbedtls/core/cmac.c [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len, const mbedtls_cipher_id_t cipher_id)
+{
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:105]: Remove blank line

File: lib/libmbedtls/core/cmac.c [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+	uint8_t block_size = cipher_id == MBEDTLS_CIPHER_ID_AES ? TEE_AES_BLOCK_SIZE : TEE_DES_BLOCK_SIZE;
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:106]: Use `size_t`. Avoid ternary operator: initialize to `TEE_DES_BLOCK_SIZE`, then do:
```C
	if (cipher_id == MBEDTLS_CIPHER_ID_AES)
		block_size = TEE_AES_BLOCK_SIZE;
```

File: lib/libmbedtls/core/cmac.c [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+		return TEE_ERROR_BAD_STATE;
+
+	if (len < block_size)
+		memcpy(digest, tmp_digest, len);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_des3_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:129]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+{
+	return mbed_cmac_final(ctx, digest, len, MBEDTLS_CIPHER_ID_3DES);
+}
+
+static TEE_Result mbed_aes_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:135]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+{
+	return mbed_cmac_final(ctx, digest, len, MBEDTLS_CIPHER_ID_AES);
+}
+
+static void mbed_cmac_free_ctx(struct crypto_mac_ctx *ctx, const mbedtls_cipher_id_t cipher_id)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:140]: Wrap at 80 characters

File: lib/libmbedtls/core/cmac.c [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+	mbed_cmac_free_ctx(ctx, MBEDTLS_CIPHER_ID_3DES);
+}
+
+static void mbed_aes_cmac_free_ctx(struct crypto_mac_ctx *ctx)
+{
+	mbed_cmac_free_ctx(ctx, MBEDTLS_CIPHER_ID_AES);
+}
+
+static void mbed_cmac_copy_state(struct crypto_mac_ctx *dst_ctx,
+		struct crypto_mac_ctx *src_ctx, const mbedtls_cipher_id_t cipher_id)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:159]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+{
+	struct mbed_cmac_ctx *src = to_cmac_ctx(src_ctx, cipher_id);
+	struct mbed_cmac_ctx *dst = to_cmac_ctx(dst_ctx, cipher_id);
+
+	if (mbedtls_cipher_clone(&dst->cipher_ctx, &src->cipher_ctx))
+		panic();
+}
+
+static void mbed_des3_cmac_copy_state(struct crypto_mac_ctx *dst_ctx,
+		struct crypto_mac_ctx *src_ctx)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:169]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+{
+	mbed_cmac_copy_state(dst_ctx, src_ctx, MBEDTLS_CIPHER_ID_3DES);
+}
+
+static void mbed_aes_cmac_copy_state(struct crypto_mac_ctx *dst_ctx,
+		struct crypto_mac_ctx *src_ctx)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:175]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+	/*
+	 * Use a default key length for getting 'cipher_info' to do the
+	 * setup. The 'cipher_info' will need to be re-assigned with final
+	 * key length obtained in mbed_cmac_init() above.
+	 *
+	 * This is safe since 'mbedtls_cipher_base_t' (used for cipher
+	 * context) uses the same fixed allocation all key lengths.
+	 */
+	cipher_info = mbedtls_cipher_info_from_values(cipher_id,
+			key_bitlen, MBEDTLS_MODE_ECB);
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:214]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+static TEE_Result mbed_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len, const mbedtls_cipher_id_t cipher_id)
+{
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
+	uint8_t block_size = cipher_id == MBEDTLS_CIPHER_ID_AES ? TEE_AES_BLOCK_SIZE : TEE_DES_BLOCK_SIZE;
+	uint8_t block_digest[TEE_AES_BLOCK_SIZE] = { };
+
+	uint8_t *tmp_digest = NULL;
+
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:110]: Add:
```C
	COMPILE_TIME_ASSERT(TEE_AES_BLOCK_SIZE >= TEE_DES_BLOCK_SIZE);

```
Although it is quite unlikely that block size will ever change :wink: it clearly documents that the declaration of `block_digest` is correct.

File: lib/libmbedtls/core/cmac.c [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+	if (cipher_id != MBEDTLS_CIPHER_ID_3DES && cipher_id != MBEDTLS_CIPHER_ID_AES)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+
+	cipher_info = mbedtls_cipher_info_from_values(cipher_id,
+			len * 8,
+			MBEDTLS_MODE_ECB);
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:53]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+		return TEE_ERROR_BAD_STATE;
+
+	if (mbedtls_cipher_cmac_starts(&c->cipher_ctx, key, len * 8))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_des3_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:70]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+{
+	return mbed_cmac_init(ctx, key, len, MBEDTLS_CIPHER_ID_3DES);
+}
+
+static TEE_Result mbed_aes_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:76]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+{
+	return mbed_cmac_init(ctx, key, len, MBEDTLS_CIPHER_ID_AES);
+}
+
+static TEE_Result mbed_des3_cmac_update(struct crypto_mac_ctx *ctx,
+		const uint8_t *data, size_t len)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:82]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+{
+	if (mbedtls_cipher_cmac_update(&to_cmac_ctx(ctx, MBEDTLS_CIPHER_ID_3DES)->cipher_ctx,
+				data, len))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_aes_cmac_update(struct crypto_mac_ctx *ctx,
+		const uint8_t *data, size_t len)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:92]: Indentation

File: lib/libmbedtls/core/cmac.c [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+{
+	if (mbedtls_cipher_cmac_update(&to_cmac_ctx(ctx, MBEDTLS_CIPHER_ID_AES)->cipher_ctx,
+				data, len))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len, const mbedtls_cipher_id_t cipher_id)
```
Comment (jforissier) [inline lib/libmbedtls/core/cmac.c:102]: Indentation

File: lib/libmbedtls/core/cmac.c (hunk 7 above)
Comment (uglym8) [reply lib/libmbedtls/core/cmac.c:106]: I'd go with AES first as it's probably most likely scenario.

**Comment by uglym8** [comment]
@jforissier thanks a lot for your review.
I believe I addressed all your comments.

**Comment by jforissier** [comment]
> @jforissier thanks a lot for your review.
> I believe I addressed all your comments.

Thank you, the code looks good except two things:
1. All tabs have disappeared in `cmac.c` :-/ (editor settings I suppose?)
2. checkpatch (IBART CI) says "b0632b10 core: libmbedtls: CMAC: remove code duplication WARNING: Missing commit description - Add an appropriate one".

Regarding the tabs issue, I can reformat the file with one key strike in vim :) so I can fix it when merging if you prefer.

Otherwise:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jbech-linaro** [comment]
> 2\. checkpatch (IBART CI)

`s/checkpatch (IBART CI)/checkpatch (Travis CI)/`

**Comment by jforissier** [comment]
> > 2. checkpatch (IBART CI)
> 
> `s/checkpatch (IBART CI)/checkpatch (Travis CI)/`

Duh! Yes of course ;-)

**Comment by uglym8** [comment]
@jforissier sorry for formatting issue.
What is preferred vim settings here?
Mine are: ts=4 sw=4 noet

**Comment by jforissier** [comment]
> @jforissier sorry for formatting issue.
> What are preferred vim settings here?
> Mine are: `ts=4 sw=4 noet`
> thus no tab expansion.

Linux kernel style. That is `ts=8 sw=8 sts=8 noet` if I'm not mistaken. I'm using `linuxsty.vim`.

**Comment by jbech-linaro** [comment]
> if I'm not mistaken. I'm using `linuxsty.vim`.

One of the best things the OP-TEE community achieved. We managed to convert Jerome from being a QT5-editor thingy person to become a Vim-purist! 😃

**Comment by jforissier** [comment]
> > if I'm not mistaken. I'm using `linuxsty.vim`.
> 
> One of the best things the OP-TEE community achieved. We managed to convert Jerome from being a QT5-editor thingy person to become a Vim-purist! smiley

Haha that is totally true! vim and cscope! :+1: What's more, I'm also using tmux (and tmux-resurrect)...

**Comment by uglym8** [comment]
> > > if I'm not mistaken. I'm using `linuxsty.vim`.
> > 
> > 
> > One of the best things the OP-TEE community achieved. We managed to convert Jerome from being a QT5-editor thingy person to become a Vim-purist! smiley
> 
> Haha that is totally true! vim and cscope! +1 What's more, I'm also using tmux (and tmux-resurrect)...

Do not forget about ctags as well:)

**Comment by jforissier** [comment]
Ping @jenswi-linaro since you suggested this refactoring IIRC.

**Comment by uglym8** [comment]
Rebased on the most recent master.

File: lib/libmbedtls/core/cmac.c [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <utee_defines.h>
+#include <util.h>
+
+struct mbed_cmac_ctx {
+	struct crypto_mac_ctx mac_ctx;
+	mbedtls_cipher_context_t cipher_ctx;
+};
```
Comment (jenswi-linaro) [inline lib/libmbedtls/core/cmac.c:23]: This PR is  for sure an improvement, but by adding a `mbedtls_cipher_id_t cipher_id` to this struct we could take it one step further.

**Review by jenswi-linaro — COMMENTED** [review]
I'm fine with this even if we could take it one step further.
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: lib/libmbedtls/core/cmac.c (hunk 22 above)
Comment (uglym8) [reply lib/libmbedtls/core/cmac.c:23]: Yep, makes total sense.
Thanks for the suggestion.

File: lib/libmbedtls/core/cmac.c (hunk 16 above)
Comment (etienne-lms) [reply lib/libmbedtls/core/cmac.c:50]: minor: indentation still to fix (indent to opening parenthesis)

File: lib/libmbedtls/core/cmac.c (hunk 22 above)
Comment (uglym8) [reply lib/libmbedtls/core/cmac.c:24]: Done as requested.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Review by jforissier — APPROVED** [review]
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Pull request merged**